source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/readonly/ImageRenderer.java

```diff
@@ -143,56 +145,125 @@ else if (imageSize == ImageSize.Large)
+
+        // Grow wrapper layout if height is stretch (assumes the parent is a vertical LinearLayout)
+        if (image.GetHeight() == HeightType.Stretch)
+        {
+            wrapper.setLayoutParams(new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT, 1));
+        }
+        else
+        {
+            wrapper.setLayoutParams(new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT));
+        }
+
+        // ConstraintLayout requires unique id
+        if(imageView.getId() == View.NO_ID) {
+            imageView.setId(View.generateViewId());
+        }
+
         long explicitWidth = image.GetPixelWidth();
         long explicitHeight = image.GetPixelHeight();
         ImageSize imageSize = image.GetImageSize();
+        ConstraintSet imageConstraints = new ConstraintSet();
+
+        // Set horizontal alignment
+        if (image.GetHorizontalAlignment() == HorizontalAlignment.Center)
+        {
+            imageConstraints.setHorizontalBias(imageView.getId(), 0.5f);
+        }
+        else if (image.GetHorizontalAlignment() == HorizontalAlignment.Right)
+        {
+            imageConstraints.setHorizontalBias(imageView.getId(), 1);
+        }
+        else
+        {
+            imageConstraints.setHorizontalBias(imageView.getId(), 0);
+        }
 
+        // By default, scale image and maintain aspect ratio
         imageView.setAdjustViewBounds(true);
         imageView.setScaleType(ImageView.ScaleType.FIT_START);
 
-        int viewWidth = ViewGroup.LayoutParams.WRAP_CONTENT;
-        int viewHeight = ViewGroup.LayoutParams.WRAP_CONTENT;
+        // By default, constrain view to top of parent, and stretch width to fill parent
+        imageConstraints.connect(imageView.getId(), ConstraintSet.START, ConstraintSet.PARENT_ID, ConstraintSet.START);
+        imageConstraints.connect(imageView.getId(), ConstraintSet.END, ConstraintSet.PARENT_ID, ConstraintSet.END);
+        imageConstraints.connect(imageView.getId(), ConstraintSet.TOP, ConstraintSet.PARENT_ID, ConstraintSet.TOP);
+        imageConstraints.constrainWidth(imageView.getId(), ConstraintSet.MATCH_CONSTRAINT);
+        imageConstraints.constrainHeight(imageView.getId(), ConstraintSet.WRAP_CONTENT);
 
-        // explicit height and/or width given
+        // Explicit height and/or width given
         if (explicitWidth != 0 || explicitHeight != 0)
         {
             if (explicitWidth != 0 && explicitHeight != 0)
             {
+                // If both are set, ignore aspect ratio
                 imageView.setScaleType(ImageView.ScaleType.FIT_XY);
             }
             if (explicitWidth != 0)
             {
-                viewWidth = Util.dpToPixels(context, explicitWidth);
+                // Stretch width + max width = exact width
+                imageConstraints.constrainMaxWidth(imageView.getId(), Util.dpToPixels(context, explicitWidth));
```

Comment:
why is the constraint for width set with MaxWidth and height with just Height?

Reply:
Setting explicit width can cause the image to grow larger than the width of its parent (which is what happened in the issue you filed w/ large image in a small column). Setting it to stretch and then placing a maximum allows stretch as much as it can, up to the explicit width, without exceeding its parent.

For height, we never want to stretch, we either want to wrap or set fixed height.

Reply:
by "stretch" you mean the image size stretch, right? so we want to make the image as large as the container width allows but keeping the ratio so we don't  mind the height at all?

Reply:
This is when an explicit width is specified. This will effectively be like `width = Math.min(container width, explicit width)`. I will try to clarify this better in the code comment.

Aspect ratio is always maintained, unless both width and height are specified.